crates/kornia-imgproc/src/features/fast.rs

```diff
@@ -133,6 +155,236 @@ fn is_fast_corner(
+            *start_idx = ((*end_idx + 16) - *most_consecutive as usize).rem_euclid(16);
+        }
+    };
+    
+    for (idx, &pixel) in pixels.iter().enumerate() {
+        // Iterate through all pixels. Only update start and end idx when consecutive count increases.
+        if pixel > upper_threshold {
+            consecutive_brighter += 1;
+            consecutive_darker = 0;
+            
+            // If we were counting brighter pixels...
+            update_indices(&consecutive_brighter, &idx, &mut start_idx, &mut end_idx, &mut most_consecutive);
+        } else if pixel < lower_threshold {
+            consecutive_darker += 1;
+            consecutive_brighter = 0;
+            
+            // If we were counting darker pixels...
+            update_indices(&consecutive_darker, &idx, &mut start_idx, &mut end_idx, &mut most_consecutive);
+        } else {
+            update_indices(&consecutive_brighter, &idx, &mut start_idx, &mut end_idx, &mut most_consecutive);
+            update_indices(&consecutive_darker, &idx, &mut start_idx, &mut end_idx, &mut most_consecutive);
+
+            consecutive_brighter = 0;
+            consecutive_darker = 0;
+        }
+
+        if consecutive_brighter >= arc_length || consecutive_darker >= arc_length {
+            is_corner = true;
+        }
+
+        if consecutive_brighter >= 16 {
+            update_indices(&consecutive_brighter, &idx, &mut start_idx, &mut end_idx, &mut most_consecutive);
+            break;
+        } else if consecutive_darker >= 16 {
+            update_indices(&consecutive_darker, &idx, &mut start_idx, &mut end_idx, &mut most_consecutive);
+            break;
+        }
+    }
+
+    if !is_corner {
+        return (false, 0); // Not a corner found
+    }
+
+    // Sum of absolute differences for the corner score. 
+    let mut score = 0i32;
+    for offset in start_idx..start_idx + most_consecutive as usize {
+        let curr_idx = offset.rem_euclid(16); 
+        score += (center_pixel.abs_diff(*pixels[curr_idx]) - threshold) as i32;
+    }
+
+    (true, score)
+}
+
+/// Fast feature detector with Non-Maximum Suppression (NMS)
+/// Goal is to replace the previous `fast_feature_detector` implementation.
+///
+/// # Arguments
+///
+/// * `src` - The source image as Gray8 image.
+/// * `threshold` - The threshold for the fast feature detector.
+/// * `arc_length` - The total number of consecutive pixels in the Bresenham circle that must be brighter or darker than the center pixel.
+/// * `nms` - A boolean flag to enable or disable Non-Maximum Suppression.
+/// # Returns
+///
+/// A vector containing the coordinates of the detected keypoints.
+pub fn fast_feature_nms_detector(
+    src: &Image<u8, 1>,
+    threshold: u8,
+    arc_length: u8,
+    nms: bool
+) -> Result<Vec<[i32; 2]>, ImageError> {
+    let (cols, rows) = (src.cols() as i32, src.rows() as i32);
+
+    // Precompute the offsets for the Bresenham circle
+    let offsets = [
```

Comment:
One micro optimisation here could be to reduce half this array and invert just values